CAGEd-oPOSSUM version 1.0 2016/02/19
====================================

Motif enrichment analysis from CAGE-derived TSSs
------------------------------------------------
CAGEd-oPOSSUM is a web-based tool used to detect the over-representation of transcription factor binding sites (TFBS) in regions containing transcription start sites (TSS) derived from Cap Analysis of Gene Expression (CAGE) peaks. It applies the methods of over-representation analysis from oPOSSUM to CAGE derived TSSs.

The online web tool is located at <a href="http://cagedop.cmmt.ubc.ca/CAGEd_oPOSSUM">http://cagedop.cmmt.ubc.ca/CAGEd_oPOSSUM</a>. See the Help page for detailed information on using CAGEd-oPOSSUM.

Authors
-------
David Arenillas, Anthony Mathelier and Wyeth Wasserman
Centre for Molecular Medicine and Therapeutics
University of British Columbia
950 West 28th Avenue
Vancouver, BC V5Z 4H4
Canada

dave@cmmt.ubc.ca
anthony.mathelier@gmail.com
wyeth@cmmt.ubc.ca


Installation
------------
If you wish to download and install your own local copy of CAGEd-oPOSSUM, please refer to the INSTALLATION and DEPENDENCIES files.


Licence
-------
CAGEd-oPOSSUM is distributed under the terms of the GNU LGPL. Please see the included COPYING.LESSER file for more information.

For convenience some 3rd party tools have been included in the CAGEd-oPOSSUM GitHub repository, namely jQuery and jsTree.

Although not included in this GitHub repository, CAGEd-oPOSSUM also utilizes the HOMER package. HOMER is available at <a href="http://homer.salk.edu/homer">http://homer.salk.edu/homer</a>.

Please also abide by the software licencing of these packages.
