CAGEd-oPOSSUM version 1.0 2016/02/19
====================================

Motif enrichment analysis from CAGE-derived TSSs
------------------------------------------------
CAGEd-oPOSSUM is a web-based tool used to detect the over-representation of
transcription factor binding sites (TFBS) in regions containing transcription
start sites (TSS) derived from Cap Analysis of Gene Expression (CAGE) peaks.
It applies the methods of over-representation analysis from our oPOSSUM tool
to CAGE derived TSSs.

The online web tool is located at: http://cagedop.cmmt.ubc.ca/CAGEd_oPOSSUM

See the website Help page for detailed information on using the CAGEd-oPOSSUM
web tool at: http://cagedop.cmmt.ubc.ca/CAGEd_oPOSSUM/help.html


Download
--------
The CAGEd-oPOSSUM code is available for download from GitHub at:
https://github.com/wassermanlab/CAGEd-oPOSSUM

It is also available for download from the web tool itself. Please see the
Download page: http://cagedop.cmmt.ubc.ca/CAGEd_oPOSSUM/download.html


Installation
------------
If you wish to download and install your own local copy of CAGEd-oPOSSUM,
please refer to the INSTALLATION and DEPENDENCIES files.


Authors
-------
David J. Arenillas, Anthony Mathelier and Wyeth W. Wasserman
Centre for Molecular Medicine and Therapeutics
University of British Columbia
950 West 28th Avenue
Vancouver, BC V5Z 4H4
Canada

dave@cmmt.ubc.ca
anthony.mathelier@gmail.com
wyeth@cmmt.ubc.ca

Please also refer to the AUTHORS file.


Licence
-------
CAGEd-oPOSSUM is distributed under the terms of the GNU LGPL. Please see the
included COPYING.LESSER file for more information.

For convenience some 3rd party tools have been included in the CAGEd-oPOSSUM GitHub repository, namely the javascript packages jQuery, jsTree and sorttable.

Although not included in this GitHub repository, CAGEd-oPOSSUM also utilizes
the HOMER package. HOMER is available at http://homer.salk.edu/homer.

Please also abide by the software licencing of these packages.
