<html>
<head>
  <meta http-equiv="CONTENT-TYPE" content="text/html; charset=iso-8859-15">
  <link rel="stylesheet" href="styles.css" type="text/css">
  <title>FANTOM5-oPOSSUM Help</title>
</head>
<style type="text/css">
	body {
		margin:20px;
	}
	.contents_level1 {
		margin:5px;
	}
	.contents_level2 {
		margin:25px;
	}
	.contents_level3 {
		margin:45px;
	}
</style>
<body class="text">
<a name="top"></a>
  <h1><a name="possum_help"></a>FANTOM5-oPOSSUM Help</h1>
  <!--
  <p>
  <b><center>Identification of over-represented transcription factor binding sites within CAGE peak regions</center></b>
  </p>
  -->
  <h2>Contents</h2>
  <a class="contents_level1" href="#description">Description</a><br>
  <a class="contents_level2" href="#species">Species</a><br>
  <a class="contents_level1" href="#algorithm">Basic Algorithm</a><br>
  <a class="contents_level2" href="#pre_vs_on_the_fly">Pre-computed vs. on-the-fly analysis</a><br>
  <a class="contents_level1" href="#statistical_analysis">Statistical Analysis</a><br>
  <a class="contents_level1" href="#results">Understanding the Results</a><br>
  <br>
  <a class="contents_level1" href="help/option_details.html" target="help">Detailed Description of Input Options and Results</a><br>
  <br>
  <a class="contents_level1" href="download.html" target="download">Download API</a><br>
  <br>
  <a class="contents_level1" href="faq.html" target="faq">FAQ</a><br>
   
  <p><hr noshade><br><a href="#top"><img src="images/arrow-up.gif" align="right" /></a></p>
  	
  <a name="description"></a>
  <h2>Description</h2>
  <p>
  FANTOM5-oPOSSUM is a system for determing the over-representation of transcription factor binding sites (TFBS) within regions containing CAGE peaks. The input consists of a target set of CAGE peaks, a suitable background set of CAGE peaks and the set of TFBS profiles used to search the CAGE peak sequences and related search parameters.
  </p>
  <p>The target and background CAGE peaks may come from FANTOM5 data specified either by selecting experiments with a threshold on the total number of tag counts and/or tags per million (TPM) or by specifying FANTOM5 CAGE peak IDs. A user may also enter his or her own 'custom' CAGE peaks by providing a BED file defining the CAGE peak coordinates. TFBS over-representation is calculated by comparing the number and distribution of the binding sites in the target set with the background.
  </p>
  <p>
  For each transcription factor (TF), the system uses two different statistical measures of significance of TFBS over-representation. The <a href="#zscore">Z-score</a> compares the number of binding site 'hits' in the target set against the number of hits in the background set, whereas the <a href="#fisher">Fisher score</a> compares the proportion of target sequences (CAGE peak regions) containing at least one TFBS with the proportion of background sequences containing at least one TFBS. The relative rankings of these two scoring methods can thus be used to determine which TFs are considered over-represented in the target set.
  </p>

  <a name="species"></a>
  <h3>Species</h3>
  <p>
  FANTOM5 provides expression data for <b>human</b> and <b>mouse</b>. The CAGE peak co-ordinates are based on the <b>GRCh37</b> and <b>NCBI37</b> assemblies respectively. 
  </p>

  <a name="algorithm"></a>
  <h2>Basic Algorithm</h2>
  <p>
  The user selects a target and background set of CAGE peaks by either choosing one or more FANTOM5 samples with some threshold on expression levels, inputting a list of specific FANTOM5 CAGE peak IDs or inputting a 'custom' set of CAGE peak coordinates and any applicable filters for both the target and background. TFBS search parameters are then chosen and the anlysis is launched. The following steps apply to both the target and background CAGE peaks this selected.
  </p>
  If the CAGE peaks came from FANTOM5 samples, and the user chose to filter by TSS status, then any CAGE peaks not classified as TSSs by the FANTOM5 TSS classifier are removed from the set. If the user chose to limit FANTOM5 CAGE peaks to those associated with specific genes, then any CAGE peaks which are not associated with one of the genes provided are filtered out. If the user chose to filter by genomic regions, this filter is applied later in the process as described in the next paragraph.
  <p>
  The system then applies the upstream / downstream flanking regions (chosen in the TFBS search parameters step) to each of the CAGE peaks to create initial CAGE peak regions. CAGE peak regions which overlap are merged together into larger regions. If the user chose to filter the CAGE peaks by a set of genomic regions (this option is available for both FANTOM5 CAGE peaks and user-entered 'custom' CAGE peaks), then these filtering regions are applied and only the portions of the CAGE peak regions which overlap the filtering regions are used in the anlysis, i.e. the intersection of the merged CAGE peak regions and filtering regions is used as the final set of regions to search for TFBS.
  </p>
  <p>
  The system then uses the selected TFBS profile matrices to scan these regions for putative binding sites which score above the selected score threshold. By comparing the frequency of TFBSs in the target set of CAGE peak regions to the frequency of TFBSs in the background set, a measure of the degree of over-representation of each TF's binding sites is obtained. The results display the rankings of the TFs degree of over-representation. Two different statistical tests of over-representation are applied to obtain these rankings, the <a href="#zscore">Z-score</a> and <a href="#fisher">the one-tailed Fisher exact probability</a>. This are described in more detail in the next <a href="#statistical_analysis">Statistical Analysis</a> section below.
  </p>

  <a name="pre_vs_on_the_fly"></a>
  <h3>Pre-computed vs. on-the-fly analysis</h3>
  <p>
  In the case where the user chooses CAGE peaks from FANTOM5 samples <b>and also</b> chooses JASPAR TFBS profiles, then both the merged CAGE peaks (based on the maximum allowed flanking sequence of 2000 upstream and downstream bp) and the predicted TFBS which fall within these regions have already been pre-computed and stored in a database. If the user chooses <b>either</b> custom CAGE peaks or custom TFBS profiles then the system has to do the computation on-the-fly by extracting the relevant sequences and scanning them for putative TFBS. The analysis based on pre-computed CAGE peak regions and TFBSs stored in the database will therefore generally complete much faster.
  </p>

  <a name="statistical_analysis"></a>
  <h2>Statistical Analysis</h2>
  
  <a name="zscore"></a>
  <h3>Z-score</h3>
    <p>
    The Z-score basically compares the frequency with which binding sites occur in the target set with the frequencey with which they occur in the background. As a way to allow comparison between transcription factors with differing binding profile widths the calculation is 'normalized' by comparing the relative freqencies of the total number of individual nucleotides bound rather than just the frequencies of the total number of individual sites.
    </p>
    <p>
    Mathematically, the Z-score uses the normal approximation to the binomial distribution to compare the rate of occurrence of a TFBS in the target set of genes to the expected rate estimated from the pre-computed background set.
    </p>
    <p class="text">
    For a given TFBS, let the random variable <i>X</i> denote the number of 
    <font color="blue">predicted</font> binding site nucleotides in the target set of CAGE peak regions.  Let <i>B</i> be the number of predicted binding site nucleotides comprising the <font color="blue">background</font> set of CAGE peak regions.
    </p>
    <p>Using a binomial model with <i>n</i> events, where <i>n</i> is the total number of nucleotides examined (i.e. the total number of nucleotides in the target CAGE peak regions), and <i>N</i> is the total number of nucleotides examined from the background CAGE peak regions, then the expected value of <i>X</i> is <i>u</i> = <i>B</i> * <i>C</i>, where 
    <i>C</i> = <i>n</i> / <i>N</i> (i.e. <i>C</i> is the ratio of sample sizes). 
    Then taking <i>p</i> = <i>B</i> / <i>N</i> as the probability of success, the standard deviation 
    is given by <i>s</i> = sqrt(<i>n</i> * <i>p</i> * (1 - <i>p</i>)).
    </p>
    <p>Now, let <i>x</i> be the <font color="blue">observed</font> number of binding site nucleotides in the target CAGE peak regions.  By applying the Central Limit 
    Theorem and using the normal approximation to the binomial distribution with a continuity 
    correction, the z-score is calculated as 
    <i>z</i>&nbsp;=&nbsp;(<i>x</i>&nbsp;-&nbsp;<i>u</i>&nbsp;-&nbsp;0.5)&nbsp;/&nbsp;<i>s</i>.
    </p>

  <a name="fisher"></a>
  <h3>Fisher Score</h3>
    <p>
    The Fisher score is based on one-tailed Fisher exact probability. In contrast to the z-score, for a given TF, the one-tailed Fisher exact probability compares the proportion of CAGE peak regions containing at least one predicted binding site to the proportion of the background set that contains at least one predicted binding site to determine the probability of a non-random association between the CAGE peak regions and the TF of interest.  It is calculated using the hypergeometric probability distribution that describes sampling without replacement from a finite population consisting of two types of elements. Therefore, the number of times a TFBS occurs in a specific CAGE peak region is disregarded, and instead, the TFBS is considered as either present or absent. Fisher exact probabilities were calculated using the R Statistics package (http://www.r-project.org/). Negative natural logarithms of the probabilities are used as the Fisher scores.
    </p>

  <a name="results"></a>
  <h2>Understanding the Results</h2>
  <p>
  In general, the scores are used to *rank* the over-representation of a TFs putative
  binding sites from the most strongly over-represented to the least, to aid you
  in selecting potential TFs of interest. The scores are
  dependent upon a number of factors, one of which is the number of CAGE peak regions
  analyzed, thus comparing scores between different analyses should not be done
  unless the number and length of regions used in the analyses are similar. Another factor
  is your selection of background. If your background does not have a similar
  nucleotide composition as the target set of CAGE peak regions, you risk biasing your analysis
  to a subset of TF motifs. To detect whether a bias has occurred a simple visualization is to plot the Z-/Fisher score (y-axis) against the TF profile's GC content (x-axis). The system will automatically generate these plots for you.
  If you see a skew (obvious slope to the distribution of scores in the plot) such that the TF matrices with a high (or low) GC content to all have the highest ranking scores, then you may need to go back and chose a different background set. 
  </p>
  <p>
  There is no specific threshold that can be recommended for any one data set;
  however within the results of an analysis you may be able to select a group of
  interesting TFs by the relative value of the scores. If your experiments were
  designed to provide well-defined data sets highlighting genes regulated by a TF,
  then it is common to see one or several TFs with scores clearly higher than the
  majority of the scores. In general a good overview of your results can be obtained
  by graphical plot. For instance you might plot the TFs by rank on the x-axis and
  a corresponding score on the y-axis, or you might plot both scores against each
  other, or you can use the plot mentionned in the prior paragraph to display one
  of the scores vs. the GC content of the TF profiles.  In our experience, a clear
  segregation of scores is the most reliable indication of functional relevance.
  </p>
	
<!--
  <ul>
    <li><h4><a href="help/help_ssa.html" name="ssa">Single Site Analysis (SSA)</a></h4>
    <li><h4><a href="help/help_acsa.html" name="acsa">Anchored Combination Site Analysis (aCSA)</a></h4>
    <li><h4><a href="help/help_tca.html" name="tca">TFBS Cluster Analysis (TCA)</a></h4>
    <li><h4><a href="help/help_actca.html" name="actca">Anchored Combination TFBS Cluster Analysis (aCTCA)</a></h4>
  </ul>
-->
