<html>
<head>
  <meta charset="UTF-8">
  <link rel="stylesheet" href="styles.css" type="text/css">
  <title>CAGEd-oPOSSUM Help</title>
</head>
<style type="text/css">
	body {
		margin:20px;
	}
	.contents_level1 {
    font-size: 20px;
		margin:5px;
	}
	.contents_level2 {
		margin:25px;
	}
	.contents_level3 {
		margin:45px;
	}
</style>

<body class="text">
<a name="top"></a>
  <h1><a name="possum_help"></a>CAGEd-oPOSSUM Help</h1>
  <!--
  <p>
  <b><center>Identification of over-represented transcription factor binding sites within CAGE peak regions</center></b>
  </p>
  -->
  <h2>Contents</h2>
  <a class="contents_level1" href="#overview">Overview</a><br>
  <br>
  <a class="contents_level1" href="#species">Species and Assemblies</a><br>
  <br>
  <a class="contents_level1" href="#description">General Description</a><br>
  <br>
  <a class="contents_level1" href="#algorithm">Basic Algorithm</a><br>
  <a class="contents_level2" href="#pre_vs_on_the_fly">Pre-computed vs. on-the-fly analysis</a><br>
  <br>
  <a class="contents_level1" href="#statistical_analysis">Statistical Analysis</a><br>
  <a class="contents_level2" href="#zscore">Z-score</a><br>
  <a class="contents_level2" href="#fisher">Fisher score</a><br>
  <br>
  <a class="contents_level1" href="#results">Understanding the Results</a><br>
  <br>
  <a class="contents_level1" href="#detailed_description">Detailed Description of Input Options and Results</a><br>
  <br>
  <a class="contents_level1" href="#download">Download API</a><br>
  <br>
  <a class="contents_level1" href="#faq">FAQ</a><br>
  <br>
  <a class="contents_level1" href="#citing">Citing CAGEd-oPOSSUM</a><br>
   
  <p><hr noshade><br><a href="#top"><img src="images/arrow-up.gif" align="right" /></a></p>
  	
  <a name="overview"></a>
  <h2>Overview</h2>
  <p>
  CAGEd-oPOSSUM is a web-based tool which may be used to determine the over-representation of transcription factor binding sites (TFBS) within CAGE peak regions. The input consists of a foreground (target) set of CAGE peaks, a suitable background set of CAGE peaks (or compositionally matched random genomic background) and a set of TFBS profiles. Optional filters may be applied to both the target and background CAGE peaks and TFBS search parameters specified. CAGEd-oPOSSUM then compares the frequency of binding sites for each transcription factor (TF) in the target versus the background and the degree of over-representation is measured statistically. 
  </p>

  <a name="species"></a>
  <h2>Species / Assemblies</h2>
  <p>
  FANTOM5 provides expression data for <b>human</b> and <b>mouse</b>. The CAGE peak co-ordinates are based on the human <b>GRCh37/hg19</b> and mouse <b>NCBI37/mm9</b> assemblies. 
  </p>

  <a name="description"></a>
  <h2>General Description</h2>
  <p>
  The target and background CAGE peaks may either be selected from the FANTOM5 CAGE peak data or supplied as your own "custom" CAGE peaks from some other source. If FANTOM5 CAGE peak data is used, it may be specified by either selecting one or more samples from the FANTOM5 sample ontology tree along with a minimum level of expression of those CAGE peaks within the selected samples or by providing a list of specific FANTOM5 CAGE peak IDs. If custom CAGE peaks are provided, the CAGE peak coordinates are specified in BED format. The background may also be generated as a random set of genomic sequences which are %GC composition and length matched to the target set using the <a href="http://homer.salk.edu/homer" target=_blank>HOMER software</a>.
  </p>
  Varous optonal filters may be provided. For FANTOM5 CAGE peaks, these filters include (1) limiting CAGE peaks to only those classified as TSSs by the FANTOM5 <a href="http://fantom.gsc.riken.jp/5/datafiles/phase1.3/extra/TSS_classifier/TSSpredictionREADME.pdf" target=_blank>TSS classifier</a>, (2) limiting CAGE peaks to only those associated to genes in a provided list of gene identifiers, or (3) limiting the CAGE peak TFBS search regions to only those portions which overlap a provided set of filtering regions. If you use your own CAGE peaks, then only the filtering by a set of regions option is available. If more than one filter is provided then only CAGE peaks and/or portions of the CAGE peak regions which pass ALL the filters or retained.
  <p>
  <p>
  The TFBS search parameter options include which transcription factor binding site profiles to use, what scoring threshold to apply to these binding sites, how much flanking region to apply around the CAGE peaks in which to search for TFBS and how the results will be displayed. TFBS profiles may be specified by selecting from the set of <a href="jaspar.genereg.net" target=_jaspar>JASPAR</a> CORE vertbrate profiles or by providing your own TFBS profile position frequency matrices (PFMs).
  </p>
  For each transcription factor (TF), the system uses two different statistical measures of significance of TFBS over-representation. The <a href="#zscore">Z-score</a> compares the number of binding site 'hits' in the target set against the number of hits in the background set, whereas the <a href="#fisher">Fisher score</a> compares the proportion of target sequences (CAGE peak regions) containing at least one TFBS with the proportion of background sequences containing at least one TFBS. The relative rankings of these two scoring methods can thus be used to determine which TFs are considered over-represented in the target set.
  </p>
  <p>
  For a detailed explanation of the various input options and output results formats, please see the <a href="help/option_details.html" target="help_detail">Detailed Description of Input Options and Results</a> page.
  </p>

  <a name="algorithm"></a>
  <h2>Basic Algorithm</h2>
  <p>
  Once you have selected target and background CAGE peaks (or ranomly generated %GC composition and length matched regions), any applicable target or background filters and TFBS search parameters, the analysis is launched and the following steps take place.
  </p>
  Any applicable filters you chose are applied. Filters may have been provided for either or both the target and background CAGE peaks. The one exception is, that in the case of a randomly generated %GC composition and length matched background, no filters are applicable. If you chose to filter the CAGE peaks by TSS status, then any CAGE peaks not classified as TSSs by the FANTOM5 TSS classifier are removed from the set. If you chose to limit CAGE peaks to those associated with specific genes, then any CAGE peaks which are not associated with any of the genes provided are filtered out. NOTE: these two filters only apply if you chose FANTOM5 CAGE peaks in the first step. Filtering by genomic regions applies to both FANTOM5 CAGE peaks and your own "custom" CAGE peaks. This filter is applied later in the process as described in the next paragraph.
  <p>
  CAGEd-oPOSSUM then applies the upstream / downstream flanking regions (chosen in the TFBS search parameters step) to each of the CAGE peaks to create initial CAGE peak regions. CAGE peak regions which overlap are merged together into larger regions. If you chose to filter the CAGE peaks by a set of genomic regions (this option is available for both FANTOM5 CAGE peaks and your own 'custom' CAGE peaks), then these filtering regions are applied and only the portions of the CAGE peak regions which overlap the filtering regions are used in the anlysis, i.e. the intersection of the merged CAGE peak regions and the filtering regions is used as the final set of regions to search for TFBS.
  </p>
  <p>
  CAGEd-oPOSSUM then uses the selected TFBS profile matrices to scan these regions for putative binding sites which score above the selected score threshold. By comparing the frequency of TFBS in the target set of CAGE peak regions to the frequency of TFBS in the background set, a measure of the degree of over-representation of each TF's binding sites is obtained. The results display the rankings of the TFs degree of over-representation. Two different statistical tests of over-representation are applied to obtain these rankings, the <a href="#zscore">Z-score</a> and <a href="#fisher">the one-tailed Fisher exact probability</a>. This are described in more detail in the <a href="#statistical_analysis">Statistical Analysis</a> section below.
  </p>

  <a name="pre_vs_on_the_fly"></a>
  <h3>Pre-computed vs. on-the-fly analysis</h3>
  <p>
  A pre-computation was performed in which flanking regions of 2000 bp (the maximum allowed in the analysis) were applied both up- and downstream of each FANTOM5 CAGE peak. Any overlapping regions which resulted were merged to form a set of maximally spanning non-overlapping regions. The sequences corresponding to these regions were retrieved and scanned with all JASPAR CORE vertebrate TFBS profiles which have an information content (specificity) of at least 8 bits (the minimum allowed in the analysis). The predicted binding sites which had a relative motif score of at least 80% (the minimum allowed in the analysis) were retained and all maximally spanning regions and predicted binding sites were stored in the CAGEd-oPOSSUM database.
  </p>
  <p>
  For analyses in which you select FANTOM5 CAGE peaks <b>and</b> also choose JASPAR TFBS profiles, the pre-computed TFBS are retrieved directly from the database. For analyses in which you use your own custom CAGE peaks or custom TFBS profiles <b>or</b> in which you use a randomly generated compositionally matched background, CAGEd-oPOSSUM does the computation described above "on-the-fly". Note that the target and background TFBS are computed independently, so it is quite possible that the target set is retrieved from the pre-computed database and the background is computed on-the-fly (or vice versa). Analyses that are able to retrieve pre-computed TFBS stored in the database will generally complete more quickly.
  </p>

  <a name="statistical_analysis"></a>
  <h2>Statistical Analysis</h2>
  
  <a name="zscore"></a>
  <h3>Z-score</h3>
    <p>
    The Z-score compares the frequency with which binding sites occur in the target set with the frequencey with which they occur in the background. As a way to allow comparison between transcription factors with differing binding profile widths, the calculation is "normalized" by comparing the frequency of <b>nucleotides which comprise the binding sites</b> rather than just comparing the frequencies of the <b>binding sites</b> themselves.
    </p>
    <p>
    Mathematically, the Z-score uses the normal approximation to the binomial distribution to compare the rate of occurrence of TFBS nucleotides in the target set of CAGE peak regions to the expected rate estimated from the background.
    </p>
    <p class="text">
    For a given TFBS, let the random variable <b><em>X</em></b> denote the number of 
    <font color="blue">predicted</font> binding site nucleotides in the target set of CAGE peak regions.  Let <b><em>B</em></b> be the number of predicted binding site nucleotides comprising the <font color="blue">background</font> set of CAGE peak regions.
    </p>
    <p>Using a binomial model with <b><em>n</em></b> events, where <b><em>n</em></b> is the total number of nucleotides examined (i.e. the total number of nucleotides in the target CAGE peak regions), and <b><em>N</em></b> is the total number of nucleotides examined from the background CAGE peak regions, then the expected value of <b><em>X</em></b> is <b><em>&mu;</em></b> = <b><em>B</em></b> x <b><em>C</em></b>, where 
    <b><em>C</em></b> = <b><em>n</em></b> / <b><em>N</em></b> (i.e. <b><em>C</em></b> is the ratio of sample sizes). 
    Then taking <b><em>P</em></b> = <b><em>B</em></b> / <b><em>N</em></b> as the probability of success, the standard deviation 
    is given by:
    <br>
    <br>
    <b><em>&sigma;</em></b> = sqrt(<b><em>n</em></b> x <b><em>P</em></b> x (1 - <b><em>P</em></b>)).
    <br>
    </p>
    <p>Now, let <b><em>x</em></b> be the <font color="blue">observed</font> number of binding site nucleotides in the target CAGE peak regions. By applying the Central Limit 
    Theorem and using the normal approximation to the binomial distribution with a continuity 
    correction, the z-score is calculated as:
    <br>
    <br>
    <b><em>z</em></b>&nbsp;=&nbsp;(<b><em>x</em></b>&nbsp;-&nbsp;<b><em>&mu;</em></b>&nbsp;-&nbsp;0.5)&nbsp;/&nbsp;<b><em>&sigma;</em></b>.
    </p>
    <br>

  <a name="fisher"></a>
  <h3>Fisher score</h3>
    <p>
    The Fisher score is based on one-tailed Fisher exact probability. In contrast to the z-score, for a given TF, the one-tailed Fisher exact probability compares the proportion of CAGE peak regions containing at least one predicted binding site to the proportion of the background set that contains at least one predicted binding site to determine the probability of a non-random association between the CAGE peak regions and the TF of interest.  It is calculated using the hypergeometric probability distribution that describes sampling without replacement from a finite population consisting of two types of elements. Therefore, the number of times a TFBS occurs in a specific CAGE peak region is disregarded, and instead, the TFBS is considered as either present or absent. Fisher exact probabilities were calculated using the R Statistics package (http://www.r-project.org/). Negative natural logarithms of the probabilities are used as the Fisher scores.
    </p>

  <a name="results"></a>
  <h2>Understanding the Results</h2>
  <p>
  In general, the scores are used to <em>rank</em> the over-representation of a TFs putative
  binding sites from the most strongly over-represented to the least, to aid you
  in selecting potential TFs of interest. The scores are
  dependent upon a number of factors, one of which is the number of CAGE peak regions
  analyzed, thus comparing scores between different analyses should not be done
  unless the number and length of regions used in the analyses are similar. Another factor
  is your selection of background. If your background does not have a similar
  nucleotide composition as the target set of CAGE peak regions, you risk biasing your analysis
  to a subset of TF motifs. To detect whether a bias has occurred a simple visualization is to plot the Z-/Fisher score (y-axis) against the TF profile's GC content (x-axis). The system will automatically generate these plots for you.
  If you see a skew (obvious slope to the distribution of scores in the plot) such that the TF matrices with a high (or low) GC content to all have the highest ranking scores, then you may need to go back and chose a different background set. 
  </p>
  <p>
  There is no specific threshold that can be recommended for any one data set;
  however within the results of an analysis you may be able to select a group of
  interesting TFs by the relative value of the scores.
  In general a good overview of your results can be obtained by the graphical
  plots of Z- or Fisher score versus the %GC composition of the TFBS profiles.
  In our experience, a clear segregation of scores is the most reliable indication of functional relevance.
  </p>

  <a name="detailed_description"></a>
  <h2>Detailed Description of Input Options and Results</h2>
  <p>
  For a much more detailed description of the input options and results please see the <a href="help/option_details.html" target="help_detail">Detailed Description of Input Options and Results</a> page.
  </p>

  <a name="download">Download API</a>
  <h2>Download API</h2>
  <p>
  To download the CAGEd-oPOSSUM API please see the <a href="download.html" target="download">Download API</a> page.
  </p>

  <a name="faq"></a>
  <h2>FAQ</h2>
  <p>
  Frequently asked questions page is here: <a href="faq.html" target="faq">FAQ</a>
  </p>


  <a name="citing"></a>
  <h2>Citing CAGEd-oPOSSUM</h2>
  <p>
  If you use CAGEd-oPOSSUM in your work, please cite:
  <br>
  PLACEHOLDER, UPDATE WHEN PUBLICATION ACCEPTED
  <br>
  David J. Arenillas, Erik Arner, Christopher J. Mungall, Terrence F. Meehan, Alexander D. Diehl, Hideya Kawaji, other F5 members?, The FANTOM Consortium, Yoshihide Hayashizaki, Alistair R.R. Forrest, Wyeth W. Wasserman and Anthony Mathelier
  </p>

  <p>
  If as part of your analysis you also used the HOMER software, either to generate random backgrounds or to perform HOMER motif analysis, please also cite HOMER:
  <br>
  Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: <a href="http://www.ncbi.nlm.nih.gov/sites/entrez?Db=Pubmed&term=20513432[UID]" target=_blank>20513432</a>
  </p>

	
<!--
  <ul>
    <li><h4><a href="help/help_ssa.html" name="ssa">Single Site Analysis (SSA)</a></h4>
    <li><h4><a href="help/help_acsa.html" name="acsa">Anchored Combination Site Analysis (aCSA)</a></h4>
    <li><h4><a href="help/help_tca.html" name="tca">TFBS Cluster Analysis (TCA)</a></h4>
    <li><h4><a href="help/help_actca.html" name="actca">Anchored Combination TFBS Cluster Analysis (aCTCA)</a></h4>
  </ul>
-->

<br>
<br>
<br>
<br>

</body>
</html>
